JSON encoding error when encoding optional sumtypes

### Describe the bug

When a struct has a  field that is an optional sumtype, json.encode()  on the struct causes a compiler bug.

### Expected Behavior

Correctly encoding struct with field that is an optional sumtype, omitting field from encoded json string if the optional is none.

### Current Behavior

```
/tmp/v_501/optionalsumtype.11931617404690958314.tmp.c:2304:82: warning: incompatible pointer types passing '_option_main__MySumtype *' (aka 'struct _option_main__MySumtype *') to parameter of type '_option *' (aka 'struct _option *') [-Wincompatible-pointer-types]
                        _option_ok(&(main__MySumtype[]){ string_to_sumtype_main__MySumtype(&value) }, &res, sizeof(main__MySumtype));
                                                                                                      ^~~~
/tmp/v_501/optionalsumtype.11931617404690958314.tmp.c:1767:40: note: passing argument to parameter 'option' here
void _option_ok(voidptr data, _option* option, int size);
                                       ^
/tmp/v_501/optionalsumtype.11931617404690958314.tmp.c:2308:8: error: assigning to '_option_main__MySumtype' (aka 'struct _option_main__MySumtype') from incompatible type 'main__MySumtype' (aka 'struct main__MySumtype')
                        res = int_to_sumtype_main__MySumtype(&value);
                            ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/optionalsumtype.11931617404690958314.tmp.c:2313:24: warning: incompatible pointer types passing 'byte (*)[56]' (aka 'unsigned char (*)[56]') to parameter of type '_option *' (aka 'struct _option *') [-Wincompatible-pointer-types]
        _option_ok(&res.data, &ret.data, sizeof(main__MySumtype));
                              ^~~~~~~~~
/tmp/v_501/optionalsumtype.11931617404690958314.tmp.c:1767:40: note: passing argument to parameter 'option' here
void _option_ok(voidptr data, _option* option, int size);
                                       ^
2 warnings and 1 error generated.
builder error: 
==================
C error. This should never happen.
```

### Reproduction Steps

```
module main

import json

struct Test {
	optional_sumtype ?MySumtype
}

type MySumtype = string | int

fn main() {
	test := Test {}
	encoded := json.encode(test)
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

0.3.3

### Environment details (OS name and version, etc.)

V full version: V 0.3.3 f1e9a8f.adcd16b
OS: macos, macOS, 13.2.1, 22D68
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz

--- a/vlib/json/json_encode_sumtype_test.v
+++ b/vlib/json/json_encode_sumtype_test.v
@@ -1,0 +1,13 @@
+import json
+
+struct Test {
+	optional_sumtype ?MySumtype
+}
+
+type MySumtype = int | string
+
+fn test_simple() {
+	test := Test{}
+	encoded := json.encode(test)
+	assert dump(encoded) == '{}'
+}

--- a/vlib/v/gen/c/json.v
+++ b/vlib/v/gen/c/json.v
@@ -492,7 +492,11 @@ fn (mut g Gen) gen_sumtype_enc_dec(utyp ast.Type, sym ast.TypeSymbol, mut enc st
 					last_number_type = var_t
 					dec.writeln('\t\tif (cJSON_IsNumber(root)) {')
 					dec.writeln('\t\t\t${var_t} value = ${js_dec_name(var_t)}(root);')
-					dec.writeln('\t\t\t${prefix}res = ${var_t}_to_sumtype_${sym.cname}(&value);')
+					if utyp.has_flag(.option) {
+						dec.writeln('\t\t\t_option_ok(&(${sym.cname}[]){ ${var_t}_to_sumtype_${sym.cname}(&value) }, &${prefix}res, sizeof(${sym.cname}));')
+					} else {
+						dec.writeln('\t\t\t${prefix}res = ${var_t}_to_sumtype_${sym.cname}(&value);')
+					}
 					dec.writeln('\t\t}')
 				}
 			}

